Add convolutional RNN for sentence classification

daemon opened the pull request.

tuzhucheng commented on conv_rnn/README.md:
````diff
@@ -0,0 +1,18 @@
+
+```
+├── conv_rnn/
+│   ├── data/
````
Please move all data to https://github.com/castorini/data?

daemon replied:
Oh I used the script from https://github.com/castorini/Castor/tree/master/kim_cnn

tuzhucheng commented on conv_rnn/data.py:
```diff
@@ -0,0 +1,48 @@
+import numpy as np
+import os
```
Let's follow PEP8 standard and move the standard library imports before the third party imports.

tuzhucheng commented on conv_rnn/getData.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+mkdir data
```
Use `mkdir -p` so their isn't an error message if these directories already exist?

tuzhucheng commented on conv_rnn/train.py:
```diff
@@ -0,0 +1,192 @@
+    parser.add_argument("--fc_size", type=int)
+    parser.add_argument("--dropout_prob", type=float)
+    parser.add_argument("--n_feature_maps", type=float)
+    parser.add_argument("--rnn_type", type=str)
```
Use `choices` keyword argument to specify what are the possible choices - see argparse docs for details.

tuzhucheng commented on conv_rnn/train.py:
```diff
@@ -0,0 +1,192 @@
+                conv_rnn.train()
+    return conv_rnn.best_dev
+
+def do_random_search():
```
Where is this `do_random_search` and ` do_grid_search` called? Please add some driver code that trigger these functions.

tuzhucheng commented on conv_rnn/train.py:
```diff
@@ -0,0 +1,192 @@
+
+def set_seed(seed=0):
+    np.random.seed(seed)
+    torch.cuda.manual_seed(seed)
```
What happens if code is run in no cuda mode?

tuzhucheng commented on conv_rnn/test.py:
```diff
@@ -0,0 +1,38 @@
+    parser.add_argument("--data_dir", default="data", type=str)
+    args = parser.parse_args()
+
+    torch.cuda.set_device(1)
```
Expose the gpu number as a parameter?

tuzhucheng commented on conv_rnn/test.py:
```diff
@@ -0,0 +1,38 @@
+import os
+import random
+
+def set_seed(seed=0):
```
This function also appears in train.py. Let's consolidate it?

tuzhucheng commented on conv_rnn/test.py:
```diff
@@ -0,0 +1,38 @@
+    set_seed(5)
+    data_loader = data.SSTDataLoader(args.data_dir)
+    conv_rnn = torch.load(args.input_file)
+    conv_rnn.cuda()
```
Should we have a CPU-only option too?

tuzhucheng commented on conv_rnn/data.py:
```diff
@@ -0,0 +1,48 @@
+        words.append(word)
+    return words
+
+class SSTDataLoader(object):
```
Please take a look at http://pytorch.org/docs/master/data.html#torch.utils.data.Dataset and http://pytorch.org/docs/master/data.html#torch.utils.data.Dataset. These are PyTorch utility classes that can handle things like data shuffling, batching of different size easily so you don't have to manually deal with them in your code.

tuzhucheng commented on conv_rnn/model.py:
```diff
@@ -0,0 +1,129 @@
+        self.dropout = nn.Dropout(dropout)
+        self.fc1 = nn.Linear(n_fmaps + 2 * self.hidden_size, fc_size)
+        self.fc2 = nn.Linear(fc_size, config.get("n_labels", 5))
+        self.epoch = 0
```
Should epoch and best_dev be a part of the model or the training algorithm?

tuzhucheng commented on conv_rnn/model.py:
```diff
@@ -0,0 +1,129 @@
+        else:
+            rnn_seq, rnn_out = self.bi_rnn(x, h_0) # shape: (batch, 2, hidden_size)
+        rnn_out.data = rnn_out.data.permute(1, 0, 2)
+        x = self.conv(rnn_seq.unsqueeze(1)).squeeze(3) # shape: (batch, channels, seq len)
```
Nice comments of dimensions at end!

tuzhucheng commented on conv_rnn/model.py:
```diff
@@ -0,0 +1,129 @@
+        out = [t.squeeze(1) for t in rnn_out.chunk(2, 1)]
+        out.append(x)
+        x = torch.cat(out, 1).squeeze(2)
+        #x = self.dropout(x)
```
Let's avoid having commented out code

tuzhucheng commented on conv_rnn/train.py:
```diff
@@ -0,0 +1,192 @@
+    parser.add_argument("--seed", type=int)
+    parser.add_argument("--hidden_size", type=int)
+    parser.add_argument("--fc_size", type=int)
+    parser.add_argument("--dropout_prob", type=float)
```
Would it be better to move all the default values here (specify using `default` kwarg) instead of elsewhere (i.e. at the top of `train`?)

daemon commented:
Very detailed :+1:

lintool commented:
I expect nothing less!

daemon commented:
I've made all the relevant changes.

daemon commented:
Code now uses Dataset and DataLoader. An observation first though:

I noticed that I hadn't been zeroing the gradients each time before `optimizer.step()`, so I corrected it. However, the model does quite a bit worse with the fix--I tried vanilla hyperparameters too. For now, I've reverted to the incorrect zero_grad() placed in the epoch loop. 

I'll look into this later.
